### PR TITLE
Module7

### PR DESCRIPTION
**Pull request by I-Georg**
Please make sure you are raising this PR against your own repository and not the original. If it says https://github.com/CorndelWithSoftwire in the search bar right now, you are in the wrong place!

Please also don't forget to submit the PR's url to Aptem afterwards (by clicking on the corresponding exercise component, pasting the url in the textbox that appears and then pressing the finish button).

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+services:
+  - docker
+
+script:
+- echo "hello world"
```
Comment (JackMeadDev) [inline .travis.yml:5]: We can remove this step now

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+- docker build --target test --tag my-test-image .
+- docker run my-test-image todo_app/items_test.py    
+- docker run my-test-image todo_app/integration_test.py
```
Comment (JackMeadDev) [inline .travis.yml:8]: This is fine here, but in a real project we'd expect new tests to be being added regularly, so combining these in e.g. a tests folder, and then just executing all of the tests in that folder might be more manageable

File: .travis.yml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+services:
+  - docker
+
+script:
+- echo "hello world"
+- docker build --target test --tag my-test-image .
+- docker run my-test-image todo_app/items_test.py    
+- docker run my-test-image todo_app/integration_test.py  
+- docker run -e TOKEN -e KEY -e TOID -e DOINGID -e DONE my-test-image  todo_app/e2e_test.py
```
Comment (JackMeadDev) [inline .travis.yml:9]: (Similarly we might put this in an e2e_tests folder or similar)

File: Dockerfile [hunk 4]
```diff
@@ -15,28 +15,67 @@ RUN pip install "poetry==$POETRY_VERSION"
...
 
 ENTRYPOINT ["poetry", "run","flask", "run", "--host", "0.0.0.0"]
+
+# testing stage
+FROM base as test
+COPY poetry.lock pyproject.toml ./
+
+RUN poetry config virtualenvs.create false --local && poetry install --no-dev --no-root
+COPY . /todo_app 
+COPY  . /.env.test
```
Comment (JackMeadDev) [inline Dockerfile:50]: This command doesn't look right - you're copying over everything into a folder named `.env.test` - is this meant to be `COPY .env.test /.env.test`? The .env.test file is actually already copied over as part of the above line, so we can get rid of this one

File: Dockerfile [hunk 5]
```diff
@@ -15,28 +15,67 @@ RUN pip install "poetry==$POETRY_VERSION"
...
+# install firefox
+RUN FIREFOX_SETUP=firefox-setup.tar.bz2 && \
+    wget -O $FIREFOX_SETUP "https://download.mozilla.org/?product=firefox-latest&os=linux64" && \
+    tar xjf $FIREFOX_SETUP -C /opt/ && \
+    ln -s /opt/firefox/firefox /usr/bin/firefox && \
+    rm $FIREFOX_SETUP
+
+
+#RUN pip3 install selenium
+#RUN pip3 install Selenium-Screenshot
```
Comment (JackMeadDev) [inline Dockerfile:79]: We can delete these lines now 🙂

File: README.md [hunk 6]
```diff
@@ -87,5 +87,10 @@ docker build --target development --tag todo-app:dev .
 docker build --target production --tag todo-app:prod .
 docker run -p 5001:5001 --env-file .env  --mount=type=bind,source={source on your machine},target="/todo-app"  -d  todo-app
 
+docker build --target test --tag my-test-image .
```
Comment (JackMeadDev) [inline README.md:90]: A little bit of context for these would be nice, but they do work out of the box!

File: todo_app/e2e_test.py [hunk 7]
```diff
@@ -63,7 +63,26 @@ def app_with_temp_board():
 @pytest.fixture(scope="module")
 def driver():
     #with webdriver.Firefox() as driver
```
Comment (JackMeadDev) [inline todo_app/e2e_test.py:65]: We should delete old commented code

**Review by JackMeadDev — APPROVED** [review]
This looks good, I've left some optional suggestions below/in comments (and I would advise trying to make your tests less flaky), but I can see your Travis pipeline running and your tests passing, so I'm happy for you to merge this when you're ready

On the flaky tests above, I can see that the same commit both [passed](https://app.travis-ci.com/github/I-Georg/DevOps-Course-Starter/builds/232615978) and [failed](https://app.travis-ci.com/github/I-Georg/DevOps-Course-Starter/builds/233395113) in Travis - that's a good sign we've got a flaky test. I saw the same issue when running the end-to-end test locally, and I didn't dig very deep but it was significantly improved by adding a `time.sleep(1)` line into the app_with_temp_board fixture;
```python
...
import time
...

def app_with_temp_board():
    ...
    thread.start()
    time.sleep(1) # this is the new line
    yield application
    ...
```

This isn't a great solution, but nor is it a particular problem here, and might well be worth adding in your module-8 work to make those tests more reliable

File: .travis.yml (hunk 1 above)
Comment (I-Georg) [reply .travis.yml:5]: Updated,thank you

File: Dockerfile (hunk 4 above)
Comment (I-Georg) [reply Dockerfile:50]: Updated,thank you

File: Dockerfile (hunk 5 above)
Comment (I-Georg) [reply Dockerfile:79]: Updated,thank you :)

**Pull request merged**